homeassistant/components/hangouts/__init__.py

```diff
@@ -0,0 +1,350 @@
+"""
+The hangouts bot component.
+
+For more details about this platform, please refer to the documentation at
+https://home-assistant.io/components/matrix/
+"""
+import logging
+import os
+
+import voluptuous as vol
+
+import homeassistant.helpers.config_validation as cv
+from homeassistant.components.notify import (ATTR_TARGET, ATTR_MESSAGE)
+from homeassistant.const import (CONF_EMAIL,
+                                 CONF_PASSWORD,
+                                 CONF_NAME,
+                                 EVENT_HOMEASSISTANT_STOP)
+
+REQUIREMENTS = ['hangups==0.4.5']
+
+_LOGGER = logging.getLogger(__name__)
+
+SESSION_FILE = '.hangouts.conf'
+
+CONF_CONVERSATIONS = 'conversations'
+CONF_COMMANDS = 'commands'
+CONF_WORD = 'word'
+CONF_EXPRESSION = 'expression'
+
+EVENT_HANGOUTS_COMMAND = 'hangouts_command'
+
+EVENT_HANGOUTS_CONNECTED = 'hangouts_connected'
+EVENT_HANGOUTS_USERS_CHANGED = 'hangouts_users_changed'
+EVENT_HANGOUTS_CONVERSATIONS_CHANGED = 'hangouts_conversations_changed'
+
+CONF_CONVERSATION_ID = 'id'
+CONF_CONVERSATION_NAME = 'name'
+
+DOMAIN = 'hangouts'
+
+TARGETS_SCHEMA = vol.All(
+    vol.Schema({
+        vol.Exclusive(CONF_CONVERSATION_ID, 'id or name'): cv.string,
+        vol.Exclusive(CONF_CONVERSATION_NAME, 'id or name'): cv.string
+    }),
+    cv.has_at_least_one_key(CONF_CONVERSATION_ID, CONF_CONVERSATION_NAME)
+)
+MESSAGE_SEGMENT_SCHEMA = vol.Schema({
+    vol.Required('text'): cv.string,
+    vol.Optional('is_bold'): cv.boolean,
+    vol.Optional('is_italic'): cv.boolean,
+    vol.Optional('is_strikethrough'): cv.boolean,
+    vol.Optional('is_underline'): cv.boolean,
+    vol.Optional('parse_str'): cv.boolean,
+    vol.Optional('link_target'): cv.string
+})
+
+MESSAGE_SCHEMA = vol.Schema({
+    vol.Required(ATTR_TARGET): [TARGETS_SCHEMA],
+    vol.Required(ATTR_MESSAGE): [MESSAGE_SEGMENT_SCHEMA]
+})
+
+COMMAND_SCHEMA = vol.All(
+    # Basic Schema
+    vol.Schema({
+        vol.Exclusive(CONF_WORD, 'trigger'): cv.string,
+        vol.Exclusive(CONF_EXPRESSION, 'trigger'): cv.is_regex,
+        vol.Required(CONF_NAME): cv.string,
+        vol.Optional(CONF_CONVERSATIONS): [TARGETS_SCHEMA]
+    }),
+    # Make sure it's either a word or an expression command
+    cv.has_at_least_one_key(CONF_WORD, CONF_EXPRESSION)
+)
+
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: vol.Schema({
+        vol.Required(CONF_EMAIL): cv.string,
```

Comment:
It looks like username and password are only needed once to retrieve the refresh token. It uses that to retrieve a refresh token which is then used to fetch new tokens and establish the connection.

Instead of using a configuration.yaml file entry, you should use a Config Flow to create a config entry. That way you use the username/password to get the refresh token and then only store the refresh token.

Reply:
Config entries docs https://developers.home-assistant.io/docs/en/config_entries_config_flow_handler.html

Here is an example config flow to use https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/homematicip_cloud/config_flow.py#L34-L58

(note that you can just put all your logic in `async_step_user` and don't have to use the user -> init redirect)

Reply:
@balloob 
if it is correct the Hangouts component
would get an entry on the "Integrations" Setup page?
that resolves the problem with 2FA secured Accounts, i think.
With this the login data are never saved, correct? Only used to generate the token.
I'll try it, thanks 🙂